File: projects/vpn-software/src/peer.rs
//! Peer management for VPN connections

use std::net::SocketAddr;
use std::time::Instant;

use crate::crypto::CryptoState;
use crate::error::{Result, VpnError};

/// Represents the state of a VPN peer
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PeerState {
    /// Peer is disconnected
    Disconnected,
    /// Handshake in progress
    Handshaking,
    /// Peer is connected and active
    Connected,
    /// Peer connection is timing out
    TimingOut,
}

/// Represents a VPN peer
pub struct Peer {
    /// Peer's public key
    public_key: [u8; 32],
    /// Peer's endpoint address
    endpoint: Option<SocketAddr>,
    /// Allowed IP addresses
    allowed_ips: Vec<std::net::Ipv4Addr>,
    /// Current state
    state: PeerState,
    /// Cryptographic state for this peer
    crypto: CryptoState,
    /// Last activity timestamp
    last_activity: Instant,
    /// Handshake attempts
    handshake_attempts: u32,
    /// RX bytes counter
    rx_bytes: u64,
    /// TX bytes counter
    tx_bytes: u64,
}

impl Peer {
    /// Create a new peer
    pub fn new(public_key: [u8; 32], endpoint: Option<SocketAddr>) -> Self {
        Self {
            public_key,
            endpoint,
            allowed_ips: Vec::new(),
            state: PeerState::Disconnected,
            crypto: CryptoState::new(),
            last_activity: Instant::now(),
            handshake_attempts: 0,
            rx_bytes: 0,
            tx_bytes: 0,
        }
    }

    /// Get the peer's public key
    pub fn public_key(&self) -> &[u8; 32] {
        &self.public_key
    }

    /// Get the peer's endpoint
    pub fn endpoint(&self) -> Option<&SocketAddr> {
        self.endpoint.as_ref()
    }

    /// Set the peer's endpoint
    pub fn set_endpoint(&mut self, endpoint: SocketAddr) {
        self.endpoint = Some(endpoint);
        self.last_activity = Instant::now();
    }

    /// Get the peer's state
    pub fn state(&self) -> &PeerState {
        &self.state
    }

    /// Set the peer's state
    pub fn set_state(&mut self, state: PeerState) {
        self.state = state;
        self.last_activity = Instant::now();
    }

    /// Add an allowed IP address
    pub fn add_allowed_ip(&mut self, ip: std::net::Ipv4Addr) {
        self.allowed_ips.push(ip);
    }

    /// Get allowed IP addresses
    pub fn allowed_ips(&self) -> &[std::net::Ipv4Addr] {
        &self.allowed_ips
    }

    /// Check if an IP is allowed for this peer
    pub fn is_ip_allowed(&self, ip: &std::net::Ipv4Addr) -> bool {
        self.allowed_ips.contains(ip)
    }

    /// Get the crypto state
    pub fn crypto(&self) -> &CryptoState {
        &self.crypto
    }

    /// Get a mutable reference to the crypto state
    pub fn crypto_mut(&mut self) -> &mut CryptoState {
        &mut self.crypto
    }

    /// Update last activity timestamp
    pub fn update_activity(&mut self) {
        self.last_activity = Instant::now();
    }

    /// Get time since last activity
    pub fn time_since_activity(&self) -> std::time::Duration {
        self.last_activity.elapsed()
    }

    /// Increment handshake attempts
    pub fn increment_handshake_attempts(&mut self) {
        self.handshake_attempts += 1;
    }

    /// Get handshake attempts count
    pub fn handshake_attempts(&self) -> u32 {
        self.handshake_attempts
    }

    /// Reset handshake attempts
    pub fn reset_handshake_attempts(&mut self) {
        self.handshake_attempts = 0;
    }

    /// Update RX bytes
    pub fn update_rx_bytes(&mut self, bytes: u64) {
        self.rx_bytes += bytes;
        self.update_activity();
    }

    /// Update TX bytes
    pub fn update_tx_bytes(&mut self, bytes: u64) {
        self.tx_bytes += bytes;
        self.update_activity();
    }

    /// Get RX bytes
    pub fn rx_bytes(&self) -> u64 {
        self.rx_bytes
    }

    /// Get TX bytes
    pub fn tx_bytes(&self) -> u64 {
        self.tx_bytes
    }
}

/// Peer manager for handling multiple peers
pub struct PeerManager {
    /// Map of public key to peer
    peers: std::collections::HashMap<[u8; 32], Peer>,
}

impl PeerManager {
    /// Create a new peer manager
    pub fn new() -> Self {
        Self {
            peers: std::collections::HashMap::new(),
        }
    }

    /// Add a peer
    pub fn add_peer(&mut self, peer: Peer) {
        self.peers.insert(*peer.public_key(), peer);
    }

    /// Remove a peer
    pub fn remove_peer(&mut self, public_key: &[u8; 32]) -> Option<Peer> {
        self.peers.remove(public_key)
    }

    /// Get a peer by public key
    pub fn get_peer(&self, public_key: &[u8; 32]) -> Option<&Peer> {
        self.peers.get(public_key)
    }

    /// Get a mutable peer by public key
    pub fn get_peer_mut(&mut self, public_key: &[u8; 32]) -> Option<&mut Peer> {
        self.peers.get_mut(public_key)
    }

    /// Find peer by IP address
    pub fn find_peer_by_ip(&self, ip: &std::net::Ipv4Addr) -> Option<&Peer> {
        self.peers.values().find(|peer| peer.is_ip_allowed(ip))
    }

    /// Get all peers
    pub fn peers(&self) -> &std::collections::HashMap<[u8; 32], Peer> {
        &self.peers
    }

    /// Get peer count
    pub fn peer_count(&self) -> usize {
        self.peers.len()
    }

    /// Get peers that need handshake
    pub fn peers_needing_handshake(&self) -> Vec<&Peer> {
        self.peers
            .values()
            .filter(|p| *p.state() == PeerState::Disconnected || *p.state() == PeerState::TimingOut)
            .collect()
    }

    /// Update all peers (check timeouts, etc.)
    pub fn update(&mut self) {
        let timeout = std::time::Duration::from_secs(30);
        let mut to_remove = Vec::new();

        for (key, peer) in &mut self.peers {
            if peer.time_since_activity() > timeout && *peer.state() == PeerState::Connected {
                peer.set_state(PeerState::TimingOut);
            }

            if peer.time_since_activity() > timeout * 2 && *peer.state() == PeerState::TimingOut {
                to_remove.push(*key);
            }
        }

        for key in to_remove {
            self.peers.remove(&key);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::net::{Ipv4Addr, SocketAddrV4};

    #[test]
    fn test_peer_creation() {
        let public_key = [0xAA; 32];
        let endpoint = SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(192, 168, 1, 1), 51820));

        let peer = Peer::new(public_key, Some(endpoint));

        assert_eq!(peer.public_key(), &[0xAA; 32]);
        assert_eq!(peer.endpoint().unwrap(), &endpoint);
        assert_eq!(*peer.state(), PeerState::Disconnected);
    }

    #[test]
    fn test_peer_allowed_ips() {
        let mut peer = Peer::new([0xAA; 32], None);

        peer.add_allowed_ip(Ipv4Addr::new(10, 0, 0, 2));
        peer.add_allowed_ip(Ipv4Addr::new(10, 0, 0, 3));

        assert!(peer.is_ip_allowed(&Ipv4Addr::new(10, 0, 0, 2)));
        assert!(peer.is_ip_allowed(&Ipv4Addr::new(10, 0, 0, 3)));
        assert!(!peer.is_ip_allowed(&Ipv4Addr::new(10, 0, 0, 4)));
    }

    #[test]
    fn test_peer_manager() {
        let mut manager = PeerManager::new();

        let peer1 = Peer::new([0xAA; 32], None);
        let peer2 = Peer::new([0xBB; 32], None);

        manager.add_peer(peer1);
        manager.add_peer(peer2);

        assert_eq!(manager.peer_count(), 2);
        assert!(manager.get_peer(&[0xAA; 32]).is_some());
        assert!(manager.get_peer(&[0xCC; 32]).is_none());
    }

    #[test]
    fn test_find_peer_by_ip() {
        let mut manager = PeerManager::new();

        let mut peer = Peer::new([0xAA; 32], None);
        peer.add_allowed_ip(Ipv4Addr::new(10, 0, 0, 2));

        manager.add_peer(peer);

        let found = manager.find_peer_by_ip(&Ipv4Addr::new(10, 0, 0, 2));
        assert!(found.is_some());
        assert_eq!(found.unwrap().public_key(), &[0xAA; 32]);

        let not_found = manager.find_peer_by_ip(&Ipv4Addr::new(10, 0, 0, 3));
        assert!(not_found.is_none());
    }
}

File: projects/vpn-software/vpn/peer.py
"""
Peer management for VPN connections.

Tracks peer state, authentication, allowed IPs, and statistics.
"""

import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional, Dict, List

from vpn.crypto import CryptoEngine
from vpn.error import PeerError


class PeerState(Enum):
    """Connection state of a VPN peer."""
    DISCONNECTED = "disconnected"
    HANDSHAKING = "handshaking"
    AUTHENTICATING = "authenticating"
    CONNECTED = "connected"
    TIMING_OUT = "timing_out"


@dataclass
class PeerStats:
    """Traffic statistics for a peer."""
    rx_bytes: int = 0
    tx_bytes: int = 0
    rx_packets: int = 0
    tx_packets: int = 0
    connected_since: float = 0.0
    last_activity: float = 0.0


class Peer:
    """
    Represents a VPN peer connection.

    Manages cryptographic state, authentication, allowed IPs,
    and traffic statistics for a single peer.
    """

    def __init__(
        self,
        public_key: bytes,
        endpoint: Optional[tuple] = None,
        session_id: int = 0,
    ) -> None:
        self._public_key = public_key
        self._endpoint = endpoint  # (host, port)
        self._session_id = session_id
        self._state = PeerState.DISCONNECTED
        self._crypto = CryptoEngine()
        self._allowed_ips: List[str] = []
        self._stats = PeerStats()
        self._handshake_attempts = 0
        self._last_activity = time.time()
        self._authenticated = False
        self._username: Optional[str] = None

    # ---- Properties ----

    @property
    def public_key(self) -> bytes:
        return self._public_key

    @property
    def endpoint(self) -> Optional[tuple]:
        return self._endpoint

    @endpoint.setter
    def endpoint(self, value: Optional[tuple]) -> None:
        self._endpoint = value
        self._last_activity = time.time()

    @property
    def session_id(self) -> int:
        return self._session_id

    @session_id.setter
    def session_id(self, value: int) -> None:
        self._session_id = value

    @property
    def state(self) -> PeerState:
        return self._state

    @state.setter
    def state(self, value: PeerState) -> None:
        self._state = value
        self._last_activity = time.time()
        if value == PeerState.CONNECTED:
            self._stats.connected_since = time.time()

    @property
    def crypto(self) -> CryptoEngine:
        return self._crypto

    @property
    def authenticated(self) -> bool:
        return self._authenticated

    @authenticated.setter
    def authenticated(self, value: bool) -> None:
        self._authenticated = value

    @property
    def username(self) -> Optional[str]:
        return self._username

    @username.setter
    def username(self, value: Optional[str]) -> None:
        self._username = value

    @property
    def stats(self) -> PeerStats:
        return self._stats

    # ---- Allowed IPs ----

    def add_allowed_ip(self, ip: str) -> None:
        """Add an IP address that this peer is allowed to send to/receive from."""
        if ip not in self._allowed_ips:
            self._allowed_ips.append(ip)

    def remove_allowed_ip(self, ip: str) -> None:
        """Remove an allowed IP."""
        if ip in self._allowed_ips:
            self._allowed_ips.remove(ip)

    @property
    def allowed_ips(self) -> List[str]:
        return self._allowed_ips.copy()

    def is_ip_allowed(self, ip: str) -> bool:
        """Check if an IP is in the allowed list."""
        return ip in self._allowed_ips

    # ---- Handshake ----

    @property
    def handshake_attempts(self) -> int:
        return self._handshake_attempts

    def increment_handshake(self) -> None:
        self._handshake_attempts += 1
        self._last_activity = time.time()

    def reset_handshake(self) -> None:
        self._handshake_attempts = 0

    # ---- Activity / Stats ----

    def update_activity(self) -> None:
        self._last_activity = time.time()

    @property
    def time_since_activity(self) -> float:
        return time.time() - self._last_activity

    def update_rx(self, n_bytes: int) -> None:
        self._stats.rx_bytes += n_bytes
        self._stats.rx_packets += 1
        self._stats.last_activity = time.time()
        self._last_activity = time.time()

    def update_tx(self, n_bytes: int) -> None:
        self._stats.tx_bytes += n_bytes
        self._stats.tx_packets += 1
        self._stats.last_activity = time.time()
        self._last_activity = time.time()

    def __repr__(self) -> str:
        return (
            f"Peer(session={self._session_id}, state={self._state.value}, "
            f"endpoint={self._endpoint}, ips={self._allowed_ips})"
        )


class PeerManager:
    """
    Manages multiple VPN peers.

    Tracks all connected peers and provides lookup by
    public key, session ID, or IP address.
    """

    def __init__(self, timeout: float = 30.0) -> None:
        self._peers: Dict[bytes, Peer] = {}  # public_key -> Peer
        self._session_map: Dict[int, bytes] = {}  # session_id -> public_key
        self._ip_map: Dict[str, bytes] = {}  # ip -> public_key
        self._timeout = timeout

    # ---- Peer Management ----

    def add_peer(self, peer: Peer) -> None:
        """Register a new peer."""
        self._peers[peer.public_key] = peer
        if peer.session_id:
            self._session_map[peer.session_id] = peer.public_key
        for ip in peer.allowed_ips:
            self._ip_map[ip] = peer.public_key

    def remove_peer(self, public_key: bytes) -> Optional[Peer]:
        """Remove a peer."""
        peer = self._peers.pop(public_key, None)
        if peer:
            if peer.session_id in self._session_map:
                del self._session_map[peer.session_id]
            for ip in peer.allowed_ips:
                self._ip_map.pop(ip, None)
        return peer

    def get_peer_by_key(self, public_key: bytes) -> Optional[Peer]:
        return self._peers.get(public_key)

    def get_peer_by_session(self, session_id: int) -> Optional[Peer]:
        key = self._session_map.get(session_id)
        if key:
            return self._peers.get(key)
        return None

    def get_peer_by_ip(self, ip: str) -> Optional[Peer]:
        key = self._ip_map.get(ip)
        if key:
            return self._peers.get(key)
        return None

    # ---- Iteration ----

    @property
    def peer_count(self) -> int:
        return len(self._peers)

    @property
    def all_peers(self) -> list:
        return list(self._peers.values())

    # ---- Maintenance ----

    def update(self) -> List[bytes]:
        """
        Update all peers: check for timeouts.

        Returns list of public keys for peers that were removed.
        """
        to_remove: List[bytes] = []

        for key, peer in self._peers.items():
            if peer.state == PeerState.CONNECTED:
                if peer.time_since_activity > self._timeout:
                    peer.state = PeerState.TIMING_OUT
            elif peer.state == PeerState.TIMING_OUT:
                if peer.time_since_activity > self._timeout * 2:
                    to_remove.append(key)

        for key in to_remove:
            self.remove_peer(key)

        return to_remove

    def register_ip(self, ip: str, public_key: bytes) -> None:
        """Register an IP address mapping for a peer."""
        self._ip_map[ip] = public_key

    def total_traffic(self) -> tuple:
        """Get total (rx_bytes, tx_bytes) across all peers."""
        total_rx = sum(p.stats.rx_bytes for p in self._peers.values())
        total_tx = sum(p.stats.tx_bytes for p in self._peers.values())
        return total_rx, total_tx
